src/main/scala/firrtl/NewEmitter.scala

```diff
@@ -0,0 +1,344 @@
+        None
+    }
+  }
+
+  /**
+   * Checks if expr is emittable by current new emitter framework. Remove when
+   * more complete new emitter is put into place.
+   */
+  def checkExprEmittable(
+    expr: Expression
+  ): Boolean = {
+    expr match {
+      case DoPrim(op, args, consts, tpe) => {
+        op match {
+          // TODO(b/38357475): remove when twigg's inline verilog refactor is checked in
+          case Rem => false
+          case _ => {
+            args.foldLeft(true)(
+              (result, arg) => result && checkExprEmittable(arg)
+            )
+          }
+        }
+      }
+      case Reference(name, tpe) => true
+      case subField: SubField => {
+        genVerilogReferenceName(subField) match {
+          case Some(name) => true
+          case None => false
+        }
+      }
+      case subIndex: SubIndex => {
+        genVerilogReferenceName(subIndex) match {
+          case Some(name) => true
+          case None => false
+        }
+      }
+      case UIntLiteral(value, width) => true
+      case SIntLiteral(value, width) => true
+      case _ => false
+    }
+  }
+
+  def genVerilog(
+    expr: Expression,
+    resultText: StringBuilder
+  ): Unit = {
+    expr match {
+      case doprim: DoPrim => {
+        genVerilog(doprim, resultText)
+      }
+      case reference: Reference => {
+        genVerilog(reference, resultText)
+      }
+      case subField: SubField => {
+        genVerilog(subField, resultText)
+      }
+      case subIndex: SubIndex => {
+        genVerilog(subIndex, resultText)
+      }
+      case (uintLiteral: UIntLiteral) => {
+        genVerilog(uintLiteral, resultText)
+      }
+      case (sintLiteral: SIntLiteral) => {
+        genVerilog(sintLiteral, resultText)
+      }
+      case _ => {
+        throw new IllegalArgumentException("CompoundExprEmitter cannot yet handle expr: " + expr)
+      }
+    }
+  }
+
+  def genVerilog(
+    doprim: DoPrim,
+    resultText: StringBuilder
+  ): Unit = {
+    def getArg0 () : Expression = doprim.args(0)
+    def getArg1 () : Expression = doprim.args(1)
+    def getArg2 () : Expression = doprim.args(2)
+    def getConst0 () : BigInt = doprim.consts(0)
+    def getConst1 () : BigInt = doprim.consts(1)
+
+    def genBinOp (
+      arg0: Expression,
+      opString: String,
+      arg1: Expression
+    ): Unit = {
+      resultText.append("(")
+      genVerilog(arg0, resultText)
+      resultText.append(opString)
+      genVerilog(arg1, resultText)
+      resultText.append(")")
+    }
+
+    def genBinOpSigned (
+      arg0: Expression,
+      opString: String,
+      arg1: Expression
+    ): Unit = {
+      val signed = doprim.args.find(x => tpe(x).typeof[SIntType])
+      if (signed != None) {
+        tpe(arg0) match {
+          case (t:SIntType) => resultText.append("$signed(")
+          case (t:UIntType) => resultText.append("$signed({1'b0,")
+        }
+      }
+      resultText.append("(")
+      genVerilog(arg0, resultText)
```

Comment:
You are recursively calling genVerilog, which could theoretically produce incorrect code:
```
wire [15: 0] answer;
wire [15: 0] a;
wire [15: 0] a;
assign answer = (a + b) >> 1; //will not work properly, because interim result is only 16 bits.
```

This is the central problem, which is why we haven't improved the verilog emitter. You need to ensure that all intermediate results are smaller than any wire/reg referenced in the expression.